File: backend_py/app/services/imagen_service.py
import os
import requests

API_BASE = 'https://generativelanguage.googleapis.com/v1beta'
IMAGE_MODEL = os.getenv('GEMINI_IMAGE_MODEL', 'gemini-2.5-flash-image').replace('models/', '')
API_KEY = os.getenv('GOOGLE_API_KEY')


def generate_image(prompt: str) -> dict:
    if not API_KEY:
        raise ValueError('GOOGLE_API_KEY is not set')

    response = requests.post(
        f"{API_BASE}/models/{IMAGE_MODEL}:generateContent",
        params={"key": API_KEY},
        json={
            "contents": [
                {
                    "role": "user",
                    "parts": [{"text": prompt}],
                }
            ],
            "generationConfig": {"responseModalities": ["IMAGE"]},
        },
        timeout=120,
    )

    if response.status_code != 200:
        raise ValueError(f"Image generation failed: {response.status_code} {response.text}")

    data = response.json()
    parts = data.get('candidates', [{}])[0].get('content', {}).get('parts', [])
    image_part = next((part for part in parts if part.get('inlineData') or part.get('inline_data')), None)
    if not image_part:
        raise ValueError('No image data returned')

    inline = image_part.get('inlineData') or image_part.get('inline_data')
    base64_data = inline.get('data')
    mime_type = inline.get('mimeType') or inline.get('mime_type') or 'image/png'

    return {
        "dataUrl": f"data:{mime_type};base64,{base64_data}",
        "mimeType": mime_type,
    }


def batch_generate_images(image_prompts: list) -> list:
    results = []
    for prompt in image_prompts:
        try:
            image = generate_image(prompt['prompt'])
            results.append({
                **prompt,
                "success": True,
                "dataUrl": image["dataUrl"],
                "mimeType": image["mimeType"],
            })
        except Exception as exc:  # noqa: BLE001
            results.append({
                **prompt,
                "success": False,
                "error": str(exc),
            })
    return results

File: backend/src/services/imagenService.js
const IMAGE_MODEL_NAME = (process.env.GEMINI_IMAGE_MODEL || 'gemini-2.5-flash-image').replace(
  /^models\//,
  ''
);
const API_BASE = 'https://generativelanguage.googleapis.com/v1beta';

const getApiKey = () => {
  const apiKey = process.env.GOOGLE_API_KEY;
  if (!apiKey) {
    throw new Error('GOOGLE_API_KEY is not set');
  }
  return apiKey;
};

/**
 * Generate image using Gemini Imagen (when available)
 * For now, this uses a placeholder approach
 */
export const generateImage = async (prompt) => {
  try {
    const apiKey = getApiKey();
    const response = await fetch(
      `${API_BASE}/models/${IMAGE_MODEL_NAME}:generateContent?key=${apiKey}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          contents: [
            {
              role: 'user',
              parts: [{ text: prompt }],
            },
          ],
          generationConfig: {
            responseModalities: ['IMAGE'],
          },
        }),
      }
    );

    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(`Image generation failed: ${response.status} ${errorBody}`);
    }

    const data = await response.json();
    const parts = data.candidates?.[0]?.content?.parts || [];
    const imagePart = parts.find((part) => part.inlineData?.data || part.inline_data?.data);

    if (!imagePart) {
      throw new Error('No image data returned from Gemini');
    }

    const inlineData = imagePart.inlineData || imagePart.inline_data;
    const base64 = inlineData.data;
    const mimeType = inlineData.mimeType || inlineData.mime_type || 'image/png';
    const dataUrl = `data:${mimeType};base64,${base64}`;

    return {
      success: true,
      prompt,
      mimeType,
      dataUrl,
    };
  } catch (error) {
    console.error('Image generation error:', error);
    throw error;
  }
};

/**
 * Batch generate images for all slides
 */
export const batchGenerateImages = async (imagePrompts) => {
  console.log(`Generating ${imagePrompts.length} images...`);
  const results = [];

  for (const prompt of imagePrompts) {
    try {
      const result = await generateImage(prompt.prompt);
      results.push({
        ...prompt,
        success: true,
        dataUrl: result.dataUrl,
        mimeType: result.mimeType,
      });
    } catch (error) {
      results.push({
        ...prompt,
        success: false,
        error: error.message,
      });
    }
  }

  return results;
};

export default {
  generateImage,
  batchGenerateImages
};
